lib/teleterm/clusters/cluster_file_transfer.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusters
+
+import (
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/sshutils/sftp"
+	api "github.com/gravitational/teleport/lib/teleterm/api/protogen/golang/v1"
+	"github.com/gravitational/trace"
+	"io"
+	"os"
+	"time"
+)
+
+func (c *Cluster) TransferFile(request *api.FileTransferRequest, server api.TerminalService_TransferFileServer) error {
+	proxyClient, err := c.clusterClient.ConnectToProxy(server.Context())
+	if err != nil {
+		return err
+	}
+	defer proxyClient.Close()
+
+	var config *sftp.Config
+	var configErr error
+
+	if request.GetDirection() == api.FileTransferDirection_FILE_TRANSFER_DIRECTION_DOWNLOAD {
+		config, configErr = sftp.CreateDownloadConfig(request.GetSource(), request.GetDestination(), sftp.Options{})
+	} else {
+		config, configErr = sftp.CreateUploadConfig([]string{request.GetSource()}, request.GetDestination(), sftp.Options{})
+	}
+
+	if configErr != nil {
+		return trace.Wrap(configErr)
+	}
+
+	config.ProgressWriter = func(fileInfo os.FileInfo) io.Writer {
+		return newGrpcFileTransferProgress(fileInfo.Size(), server)
+	}
+
+	clusterServers, err := proxyClient.FindNodesByFilters(server.Context(), proto.ListResourcesRequest{
+		Namespace: defaults.Namespace,
+	})
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var foundServer types.Server
+	for _, clusterServer := range clusterServers {
+		if clusterServer.GetName() == request.GetServerId() {
+			foundServer = clusterServer
+			break
+		}
+	}
+
+	if foundServer == nil {
+		return trace.BadParameter("Requested server does not exist")
+	}
+
+	err = c.clusterClient.TransferFiles(server.Context(), request.GetLogin(), foundServer.GetHostname()+":0", config)
```

Comment:
I'm generally unsure about this part. 
Is there a better way to provide host address to `clusterClient.TransferFiles`? Perhaps instead of sending `serverId` from Connect I should send the hostname? By doing this I wouldn't have to connect to proxy (`clusterClient.TransferFiles` also runs this function so it takes a good few seconds all together).
Additionally, downloading a file from a leaf cluster fails with an error `No proxy address specified, missed --proxy flag?`, did I miss something obvious?

Reply:
I'll try to look into it tomorrow, I probably won't have time for that today unfortunately.

Reply:
I looked into this and I think you're right, passing the hostname seems to be the way to go. I checked it and that's how both the Web UI and tsh does it. Well, in tsh you can specify the address instead of the hostname as well.

As for the problem with leaf clusters: it looks like the server URI that is used for the `TransferFile` RPC is wrong. Mine said `/clusters/teleport-local-leaf/servers/4bbde2c9-b805-4dd8-8032-a0bed162d85b`, suggesting that `teleport-local-leaf` is a root cluster but it's a leaf cluster.

Reply:
Thanks for help with this, works perfectly!

The only thing that puzzles me is `request.GetHostname()+":0"`. 
Without a port, it fails with an error `failed connecting to node teleport_root. invalid format for proxy request: "proxy:teleport_root@default@teleport_root", expected 'proxy:host:port@cluster'`.
I assume I have to add `:0` to the hostname, so the the node address is resolved automatically?

Reply:
I don't know the details but yeah, that seems to be the case. `tsh ssh` needs to support both commands:

```
tsh scp file.text root@mbp.teleport-local:/tmp/
tsh scp -P 3022 file.text root@127.0.0.1:/tmp/
```

So I imagine in case port is set to 0, tsh tries to resolve the hostname to one of the available SSH servers.